src/debugger/Debugger.js

```diff
@@ -73,21 +75,12 @@ export class DebugServer {
   // Checking if the debugger needs to take any action on reaching this ast node
   checkForActions(ast: BabelNode) {
     if (this._checkAndUpdateLastExecuted(ast)) {
-      this.checkForBreakpoint(ast);
-      this.checkStepComplete(ast);
-    }
-  }
-
-  checkForBreakpoint(ast: BabelNode) {
-    if (this._breakpointManager.shouldStopOnBreakpoint(ast)) {
-      this.waitForRun(ast, "Breakpoint");
-    }
-  }
-
-  checkStepComplete(ast: BabelNode) {
-    let steppingType = this._stepManager.getStepperType(ast);
-    if (steppingType) {
-      this.waitForRun(ast, steppingType);
+      let stoppables: Array<StoppableObject> = this._stepManager.getAndDeleteCompletedSteppers(ast);
+      let breakpoint = this._breakpointManager.getStoppableBreakpoint(ast);
+      if (breakpoint) stoppables.push(breakpoint);
+      if (this._stopEventManager.shouldDebuggeeStop(ast, stoppables)) {
+        this.waitForRun(ast);
```

Comment:
Is this recursion necessary?

Reply:
I don't think there is a recursion here? This `waitForRun` call is inside `checkForActions`.

Reply:
I missed the collapsed code in the diff view, so never mind.